crates/ruff_linter/src/rules/refurb/rules/verbose_decimal_constructor.rs

```diff
@@ -75,22 +76,58 @@ pub(crate) fn verbose_decimal_constructor(checker: &mut Checker, call: &ast::Exp
             value: str_literal, ..
         }) => {
             // Parse the inner string as an integer.
-            let trimmed = str_literal.to_str().trim_whitespace();
-
+            //
+            // For reference, a string argument to `Decimal` is parsed in CPython
+            // using this regex:
+            // https://github.com/python/cpython/blob/ac556a2ad1213b8bb81372fe6fb762f5fcb076de/Lib/_pydecimal.py#L6060-L6077
+            // _after_ trimming whitespace from the string and removing all occurrences of "_".
+            let mut trimmed = Cow::from(str_literal.to_str().trim_whitespace());
+            if memchr::memchr(b'_', trimmed.as_bytes()).is_some() {
+                trimmed = Cow::from(trimmed.replace('_', ""));
+            }
             // Extract the unary sign, if any.
-            let (unary, rest) = if let Some(trimmed) = trimmed.strip_prefix('+') {
-                ("+", trimmed)
+            let (unary, mut rest) = if let Some(trimmed) = trimmed.strip_prefix('+') {
+                ("+", Cow::from(trimmed))
             } else if let Some(trimmed) = trimmed.strip_prefix('-') {
-                ("-", trimmed)
+                ("-", Cow::from(trimmed))
             } else {
                 ("", trimmed)
             };
+            // Extract the exponent, if any.
+            let mut e_indices = memchr::memrchr2_iter(b'e', b'E', rest.as_bytes());
+            if let Some(index) = e_indices.next() {
+                // More than one `e` is an error of type `decimal.InvalidOperation`
+                // when calling Decimal on a string. The suggested
+                // fix would turn this into a SyntaxError. To maintain
+                // the behavior of the code, we abort the check in this case.
+                if e_indices.next().is_some() {
+                    return;
+                }
+                // This range will not cause a panic: in the worst case,
+                // 'e' was the last character in `rest`, and then the right
+                // hand side will be `""`.
+                let exponent = rest[index + 1..]
+                    .strip_prefix('+')
```

Comment:
It seems that the exponential syntax also supports negative numbers:

https://github.com/astral-sh/ruff/blob/51a998306a10aab1254740f760efaecc25213d71/crates/ruff_python_parser/src/lexer.rs#L1088-L1100

Reply:
Yes, you can have negative exponents. But this lint should not suggest replacing `Decimal("1e-1")` with `Decimal(1e-1)`. They are not the same! `Decimal` offers a way to represent decimal numbers in an exact way, something that floating point representations do not support.
```py
>>>> Decimal("1e-1") == Decimal(1e-1)  # Not the same, since 0.1 can not be represented exactly as a float
False
>>>> Decimal("5e-1") == Decimal(5e-1)  # Okay, 5e-1 = 1/2 can be represented exactly
True
```

I think we need to be similarly careful with large numbers, even if they are integers:
```py
>>>> Decimal("1e22") == Decimal(1e22)
True
>>>> Decimal("1e23") == Decimal(1e23)
False
```

Reply:
Just as @sharkdp points out, the omission of negative exponents was intentional. I didn't think about large exponents though, good point!

Reply:
It's hard to figure out when this conversion is actually safe - parsing scientific notation is weird, and I can't seem to find a reference for the range where the representation is faithful. I think issues start to occur when numbers are larger than `1e16` but I'm having trouble finding a justification for that in code/documentation...

The smallest example I found so far was:

```bash
>>> Decimal(1801439850948199e1) == Decimal('1801439850948199e1')
False
```

Reply:
FURB157 should not rewrite strings to float literals. Literals with exponents are float literals, so they should not be passed to the `Decimal` constructor, or they will trigger [`decimal-from-float-literal` (RUF032)](https://docs.astral.sh/ruff/rules/decimal-from-float-literal/).

```console
$ cat ruf032.py                                                   
from decimal import Decimal
Decimal(1e16)

$ ruff check --isolated --preview --select RUF032 ruf032.py 
ruf032.py:2:9: RUF032 `Decimal()` called with float literal argument
  |
1 | from decimal import Decimal
2 | Decimal(1e16)
  |         ^^^^ RUF032
  |
  = help: Use a string literal instead

Found 1 error.
No fixes available (1 hidden fix can be enabled with the `--unsafe-fixes` option).
```

Reply:
Well that makes life easier! Thank you!